tools/n64sym.c

```diff
@@ -0,0 +1,487 @@
+        }
+    }
+}
+
+int symtable_sort_by_addr(const void *a, const void *b)
+{
+    const struct symtable_s *sa = a;
+    const struct symtable_s *sb = b;
+    // In case the address match, it means that there are multiple
+    // inlines at this address. Sort by insertion order (aka stable sort)
+    // so that we preserve the inline order.
+    if (sa->addr != sb->addr)
+        return sa->addr - sb->addr;
+    return sa->uuid - sb->uuid;
+}
+
+int symtable_sort_by_func(const void *a, const void *b)
+{
+    const struct symtable_s *sa = a;
+    const struct symtable_s *sb = b;
+    int sa_len = sa->func ? strlen(sa->func) : 0;
+    int sb_len = sb->func ? strlen(sb->func) : 0;
+    return sb_len - sa_len;
+}
+
+void process(const char *infn, const char *outfn)
+{
+    verbose("Processing: %s -> %s\n", infn, outfn);
+
+    // First, find all functions and call sites. We do this by disassembling
+    // the ELF file and grepping it.
+    elf_find_callsites(infn);
+    verbose("Found %d callsites\n", stbds_arrlen(symtable));
+
+    // Compact the file names to avoid common prefixes
+    // FIXME: we need to improve this to handle multiple common prefixes
+    // eg: /home/foo vs /opt/n64/include
+    //compact_filenames();
+
+    // Sort the symbole table by symbol length. We want longer symbols
+    // to go in first, so that shorter symbols can be found as substrings.
+    // We sort by function name rather than file name, because we expect
+    // substrings to match more in functions.
+    verbose("Sorting symbol table...\n");
+    qsort(symtable, stbds_arrlen(symtable), sizeof(struct symtable_s), symtable_sort_by_func);
+
+    // Go through the symbol table and build the string table
+    verbose("Creating string table...\n");
+    for (int i=0; i < stbds_arrlen(symtable); i++) {
+        if (i % 5000 == 0)
+            verbose("  %d/%d\n", i, stbds_arrlen(symtable));
+        struct symtable_s *sym = &symtable[i];
+        if (sym->func)
+            sym->func_sidx = stringtable_add(sym->func);
+        else
+            sym->func_sidx = -1;
+        if (sym->file)
+            sym->file_sidx = stringtable_add(sym->file);
+        else
+            sym->file_sidx = -1;
+    }
+
+    // Sort the symbol table by address
+    qsort(symtable, stbds_arrlen(symtable), sizeof(struct symtable_s), symtable_sort_by_addr);
+
+    // Fill in the function offset field in the entries in the symbol table.
+    verbose("Computing function offsets...\n");
+    compute_function_offsets();
+
+    // Write the symbol table to file
+    verbose("Writing %s\n", outfn);
+    FILE *out = fopen(outfn, "wb");
+    if (!out) {
+        fprintf(stderr, "Cannot create file: symtable.bin\n");
+        exit(1);
+    }
+
+    fwrite("SYMT", 4, 1, out);
+    w32(out, 2); // Version
+    int addrtable_off = w32_placeholder(out);
+    w32(out, stbds_arrlen(symtable));
+    int symtable_off = w32_placeholder(out);
+    w32(out, stbds_arrlen(symtable));
```

Comment:
So basically the address table contains the address for each symbol in the symbol table.

Looking at the code right now, there's no big win anymore in keeping the address separated from the symbol itself. It used to be more useful for alignment / lookups / dmas reasons but I don't see such a strong benefit anymore. We might evaluate merging the address table and the symbol table to simplify a bit readability of the code.

Reply:
I don't remember exactly, the code underwent two or three refactorings. Anyway, it could be further improved now but I don't have a rush to do it, as it works quite well.